// @sitecore/byoc/src/lib/types.ts
/**
 * The DataSettings function definition is part of the @sitecore-feaas/clientside package.
 * This is the return type of the DataSettings function copied from the package to avoid a circular dependency.
 *
 * @located in  https://github.com/Sitecore-PD/sitecore.components.monorepo/blob/main/clientside/src/utils/settings.ts
 */
export type DataSettings = {
  params: Record<string, any>
  headers: Record<string, string>
  jsonpath: string
  method: 'HEAD' | 'GET' | 'POST' | 'OPTIONS' | 'PUT' | 'DELETE' | 'UPDATE'
  body: Record<string, any>
  url: string
}

/**
 * This is the DataSource type copied from the @sitecore-feaas/sdk package for reference.
 *
 * @located in  https://github.com/Sitecore-PD/sitecore.components.monorepo/blob/main/sdk/src/models/datasources.ts
 */

/** Defines the schema for Datasource. */
export interface Datasource {
  /** The identifier of the Library that Datasource belongs to */
  libraryId: string

  /** The identifier of the Datasource */
  id: string

  /**
   * The name of the Datasource.
   *
   * @minLength 1
   * @maxLength 128
   */
  name: string

  /**
   * The description of the Datasource.
   *
   * @minLength 0
   * @maxLength 128
   */
  description: string

  /** A sample of the Datasource */
  sample: object[] | object

  /** The timestamp the Datascource was created at */
  createdAt?: Date

  /** The timestamp when the Datascource was last modified at */
  modifiedAt?: Date

  /**
   * The timestamp when the Datascource was last sampled at.
   *
   * @nullable
   */
  sampledAt?: Date

  /**
   * Definition is a data returned from external resource, that we use to extract sample data from. Having this data, we
   * can detect if upstream structure was updated.
   *
   * @nullable
   */
  definition?: object | object[]

  /**
   * Definition of remote datasource for a conflicting datasource.
   *
   * @nullable
   */
  conflictDefinition?: object | object[]

  /**
   * Datasource type (internal, xm template, etc.)
   *
   * @nullable
   */
  type?: 'internal' | 'external' | 'xmTemplate' | 'contentHubOne'

  /**
   * Id of external source. Eg: Tenant name for XM Cloud datasources, Project Id for internal datasources etc.
   *
   * @nullable
   */
  externalSourceId?: string

  settings: DataSettings
}
