// @sitecore-jss/sitecore-jss-react/dist/cjs/components/MissingComponent.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MissingComponent = void 0;
const react_1 = __importDefault(require("react"));
const MissingComponent = (props) => {
    const componentName = props.rendering && props.rendering.componentName
        ? props.rendering.componentName
        : 'Unnamed Component';
    // error override would mean component is not unimplemented
    !props.errorOverride &&
        console.log(`Component props for unimplemented '${componentName}' component`, props);
    const errorMessage = props.errorOverride ||
        'JSS component is missing React implementation. See the developer console for more information.';
    return (react_1.default.createElement("div", { style: {
            background: 'darkorange',
            outline: '5px solid orange',
            padding: '10px',
            color: 'white',
            maxWidth: '500px',
        } },
        react_1.default.createElement("h2", null, componentName),
        react_1.default.createElement("p", null, errorMessage)));
};
exports.MissingComponent = MissingComponent;
exports.MissingComponent.displayName = 'MissingComponent';
